flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/FlinkSink.java

```diff
@@ -0,0 +1,635 @@
+      // schema.
+      return (RowType) requestedSchema.toRowDataType().getLogicalType();
+    } else {
+      return FlinkSchemaUtil.convert(schema);
+    }
+  }
+
+  static DataStream<RowData> distributeDataStream(DataStream<RowData> input,
+                                                  Map<String, String> properties,
+                                                  List<Integer> equalityFieldIds,
+                                                  PartitionSpec partitionSpec,
+                                                  Schema iSchema,
+                                                  RowType flinkRowType,
+                                                  DistributionMode distributionMode,
+                                                  List<String> equalityFieldColumns) {
+    DistributionMode writeMode;
+    if (distributionMode == null) {
+      // Fallback to use distribution mode parsed from table properties if don't specify in job level.
+      String modeName = PropertyUtil.propertyAsString(
+          properties,
+          WRITE_DISTRIBUTION_MODE,
+          WRITE_DISTRIBUTION_MODE_NONE);
+
+      writeMode = DistributionMode.fromName(modeName);
+    } else {
+      writeMode = distributionMode;
+    }
+
+    LOG.info("Write distribution mode is '{}'", writeMode.modeName());
+    switch (writeMode) {
+      case NONE:
+        if (equalityFieldIds.isEmpty()) {
+          return input;
+        } else {
+          LOG.info("Distribute rows by equality fields, because there are equality fields set");
+          return input.keyBy(new EqualityFieldKeySelector(iSchema, flinkRowType, equalityFieldIds));
+        }
+
+      case HASH:
+        if (equalityFieldIds.isEmpty()) {
+          if (partitionSpec.isUnpartitioned()) {
+            LOG.warn("Fallback to use 'none' distribution mode, because there are no equality fields set " +
+                "and table is unpartitioned");
+            return input;
+          } else {
+            return input.keyBy(new PartitionKeySelector(partitionSpec, iSchema, flinkRowType));
+          }
+        } else {
+          if (partitionSpec.isUnpartitioned()) {
+            LOG.info("Distribute rows by equality fields, because there are equality fields set " +
+                "and table is unpartitioned");
+            return input.keyBy(new EqualityFieldKeySelector(iSchema, flinkRowType, equalityFieldIds));
+          } else {
+            for (PartitionField partitionField : partitionSpec.fields()) {
+              Preconditions.checkState(equalityFieldIds.contains(partitionField.sourceId()),
+                  "In 'hash' distribution mode with equality fields set, partition field '%s' " +
+                      "should be included in equality fields: '%s'", partitionField, equalityFieldColumns);
+            }
+            return input.keyBy(new PartitionKeySelector(partitionSpec, iSchema, flinkRowType));
```

Comment:
Since the partition fields are included in equality fields, should we use `EqualityFieldKeySelector`?

Reply:
Good question. When we have more equality fields than the partition fields, will it cause data distribution errors if we use 
 EqualityFieldKeySelector?

Reply:
Why do we need `EqualityFieldKeySelector`? Will the requirement be lost if we use `PartitionKeyFieldSelector`?  Do we need another `KeyBy` operator to handle the case that equality keys are not the source of partition keys?

Reply:
If equality keys are not the source of partition keys, then the data with the same equal column value may be divided into different partitions, which is not allowed.
use PartitionKeyFieldSelector, the data is written to the correct partition, and because the partitioned fields is part of the equal fields, it does not cause data results to be abnormal.

Reply:
> If equality keys are not the source of partition keys, then the data with the same equal column value may be divided into different partitions, which is not allowed. 

The check above states that _"In 'hash' distribution mode with equality fields set, partition field '%s' " +
                      "should be included in equality fields: '%s'"_. It contradicts what you said.

Reply:
For example, if we have a composite key `(id, grade)` and a partition field `(grade)`, the `PartitionKeyFieldSelector` distribute data according to `grade` field which is apparently not enough. Or do you mean that we should build `PartitionSpec` upon both `id` and `grade`?

Reply:
> > If equality keys are not the source of partition keys, then the data with the same equal column value may be divided into different partitions, which is not allowed.
> 
> The check above states that _"In 'hash' distribution mode with equality fields set, partition field '%s' " + "should be included in equality fields: '%s'"_. It contradicts what you said.

Sorry for the confusion, I got sidetracked.

The reason we used  EqualityFieldKeySelector, because we need to ensure all the same primary key record will be distributed to the same IcebergStreamWriter to ensure result correctness, . Bug when users set HASH distribution, the intention is to cluster data by partition columns. If the partition distribution satisfies the equality distribution, then this should use the partition distribution. Otherwise, it should use the equality distribution to guarantee correctness. The key point here, then, is that when the partition distribution is equal, we can use the partition distribution without violating the user's intent. So what is `satisfies the equality distribution`? Requiring all of the partition source fields are identifier fields. 
refer to: #2898 https://github.com/apache/iceberg/pull/2898#discussion_r810411948

Reply:
I see. Thanks for the explanation!